### DISCUSSION TITLE
In space and time models, how to include time-varying covariates in lgcp inlabru

### DISCUSSION THREAD

**Discussion by bacuu1** [inlabru-org/inlabru#234] [Q&A]
Hello team inlabru

This is from the vignettes on the inlabru website, an example in space and time. 
How can we include a time-varying covariate in such models? 
```r
matern <- inla.spde2.pcmatern(mrsea$mesh,
  prior.sigma = c(0.1, 0.01),
  prior.range = c(10, 0.01)
)

cmp <- coordinates + season ~ Intercept(1) +
  mySmooth(
    coordinates,
    model = matern,
    group = season,
    ngroup = 4
  ) 

fit <- lgcp(cmp,
  data = mrsea$points,
  samplers = mrsea$samplers,
  domain = list(
    coordinates = mrsea$mesh,
    season = seq_len(4)
  )
)
```

**Comment by finnlindgren** [comment]
Sorry about the slow reply!
For spatially varying covariates that are constant in time, the following are the main options:
```r
# Assuming `covar` is a gridded covariate in `terra` `SpatRaster` format:
# Automatic extraction:
comp1 <- ~ 0 + covariate(covar) + ...
# Explicit extraction using the default extractor method:
comp2 <- ~ 0 + covariate(eval_spatial(covar, where = .data.)) + ...
# Explicit extraction using a user defined extractor method:
covar_fun <- function(data) {
  eval_spatial(covar, where = data)
}
comp3 <- ~ 0 + covariate(covar_fun(.data.)) + ...
```
The three versions above are equivalent, but the third option allows the user to write their own code for extracting the covariate values at the given locations (as given by the input data, which would typically be an `sf` object with point geometry).

To extend this to space-time, the solution depends on how the space-time covariate is stored. In the special case of storing each time as a layer in the raster:
```r
# Assuming $time in the lgcp() input data is an index variable, and domain = list(geometry=mesh, time=seq_len(n_time))
# Automatic extraction:
comp1 <- ~ 0 + covariate(covar, main_layer = time) + ...
# Explicit extraction using the default extractor method:
comp2 <- ~ 0 + covariate(eval_spatial(covar, where = .data., layer = time)) + ...
# Explicit extraction using a user defined extractor method:
covar_fun <- function(data) {
  eval_spatial(covar, where = data, selector = "time")
}
comp3 <- ~ 0 + covariate(covar_fun(.data.)) + ...
```
As before, these three options are equivalent. If the covariate information is stored differently, the user will need to write their own extraction `covar_fun()` function, supporting whatever storage method was used.
